### PR TITLE
add support for bifrost native token BNC

### PR DESCRIPTION
**Pull request by herryho**
For sake of testing, we temporarily changed the fee currency to Bifrost native token BNC.

### CODE REVIEW

File: runtime/karura/src/lib.rs [hunk 1]
```diff
@@ -1439,7 +1447,7 @@ impl xcm_executor::Config for XcmConfig {
 	type Barrier = Barrier;
 	type Weigher = FixedWeightBounds<UnitWeightCost, Call>;
 	// Only receiving KSM is handled, and all fees must be paid in KSM.
-	type Trader = FixedRateOfConcreteFungible<KsmPerSecond, ToTreasury>;
+	type Trader = FixedRateOfConcreteFungible<BifrostNativeTokenLocation, ToTreasury>;
```
Comment (shaunxw) [inline runtime/karura/src/lib.rs:1450]: This breaks KSM transfer between Karura and Kusama. Should use multiple traders instead. Blocked by https://github.com/paritytech/polkadot/pull/3601

File: primitives/src/currency.rs [hunk 2]
```diff
@@ -187,6 +187,7 @@ create_currency_id! {
 		LDOT("Liquid DOT", 10) = 3,
 		// 20 - 39: External tokens (e.g. bridged)
 		RENBTC("Ren Protocol BTC", 8) = 20,
+		BNC("Bifrost Native Token", 12) = 133,
```
Comment (shaunxw) [inline primitives/src/currency.rs:190]: `133` is reserved for Karura & Kusama native tokens. `168 - 255` is available to parachain tokens, plus pls move this line to the right section.

File: node/service/src/chain_spec/karura.rs [hunk 3]
```diff
@@ -46,7 +46,7 @@ fn karura_properties() -> Properties {
 	let mut properties = Map::new();
 	let mut token_symbol: Vec<String> = vec![];
 	let mut token_decimals: Vec<u32> = vec![];
-	[KAR, KUSD, KSM, LKSM].iter().for_each(|token| {
+	[KAR, KUSD, KSM, LKSM, BNC].iter().for_each(|token| {
```
Comment (shaunxw) [inline node/service/src/chain_spec/karura.rs:49]: not sure if BNC fits here, @xlc could you confirm?

File: runtime/karura/src/lib.rs [hunk 4]
```diff
@@ -1425,6 +1426,13 @@ impl TakeRevenue for ToTreasury {
 	}
 }
 
+parameter_types! {
+	pub BifrostNativeTokenLocation: (MultiLocation, u128) = (
+		X3(Parent, Parachain(2001), GeneralKey([0,0].to_vec())),
+		ksm_per_second()
```
Comment (shaunxw) [inline runtime/karura/src/lib.rs:1432]: `ksm_per_second` is based on KSM/KAR price ratio, which is same with BNC/KAR

**Review by shaunxw — CHANGES_REQUESTED** [review]

File: node/service/src/chain_spec/karura.rs (hunk 3 above)
Comment (xlc) [reply node/service/src/chain_spec/karura.rs:49]: I think we can have it here. Helps client to know decimals of BNC.
But also need to update https://github.com/AcalaNetwork/Acala/blob/master/resources/karura-dist.json

File: runtime/karura/src/lib.rs [hunk 5]
```diff
@@ -736,6 +736,7 @@ parameter_type_with_key! {
 				TokenSymbol::KUSD => cent(*currency_id),
 				TokenSymbol::KSM => 10 * millicent(*currency_id),
 				TokenSymbol::LKSM => 50 * millicent(*currency_id),
+				TokenSymbol::BNC => Zero::zero(),
```
Comment (xlc) [inline runtime/karura/src/lib.rs:739]: also need a non zero number here. should be something higher than 0.00001 KSM

**Comment by xlc** [comment]
Please use `cargo fmt` to format the code

**Comment by herryho** [comment]
> Please use `cargo fmt` to format the code and make sure tests are passing

Code is formatted and tests are passed. Please review.

File: runtime/karura/src/lib.rs [hunk 6]
```diff
@@ -1555,13 +1565,23 @@ impl Convert<MultiLocation, Option<CurrencyId>> for CurrencyIdConvert {
...
+						// check `currency_id` is cross-chain asset
+						match currency_id {
+							Token(KAR) | Token(KUSD) | Token(LKSM) | Token(RENBTC) => Some(currency_id),
+							_ => None,
+						}
+					} else {
+						None
+					}
+				} else if id == 2001 {
+					if key == [0, 1].to_vec() {
```
Comment (shaunxw) [inline runtime/karura/src/lib.rs:1581]:
```suggestion
				} else if id == 2001 && key == [0, 1].to_vec(){
```

**Review by shaunxw — COMMENTED** [review]
blocked by https://github.com/AcalaNetwork/Acala/pull/1391

**Review by xlc — APPROVED** [review]

**Review by shaunxw — APPROVED** [review]

**Review by xlc — APPROVED** [review]

**Pull request merged**